components/raftstore-v2/src/worker/cleanup/compact.rs

```diff
@@ -0,0 +1,365 @@
+                .field("cf_names", cf_names)
+                .field("regions", region_ids)
+                .field(
+                    "tombstones_num_threshold",
+                    &compact_threshold.tombstones_num_threshold,
+                )
+                .field(
+                    "tombstones_percent_threshold",
+                    &compact_threshold.tombstones_percent_threshold,
+                )
+                .field(
+                    "redundant_rows_threshold",
+                    &compact_threshold.redundant_rows_threshold,
+                )
+                .field(
+                    "redundant_rows_percent_threshold",
+                    &compact_threshold.redundant_rows_percent_threshold,
+                )
+                .finish(),
+        }
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("compact failed {0:?}")]
+    Other(#[from] Box<dyn StdError + Sync + Send>),
+}
+
+pub struct Runner<E> {
+    logger: Logger,
+    tablet_registry: TabletRegistry<E>,
+}
+
+impl<E> Runner<E>
+where
+    E: KvEngine,
+{
+    pub fn new(tablet_registry: TabletRegistry<E>, logger: Logger) -> Runner<E> {
+        Runner {
+            logger,
+            tablet_registry,
+        }
+    }
+}
+
+impl<E> Runnable for Runner<E>
+where
+    E: KvEngine,
+{
+    type Task = Task;
+
+    fn run(&mut self, task: Self::Task) {
+        match task {
+            Task::CheckAndCompact {
+                cf_names,
+                region_ids,
+                compact_threshold,
+            } => match collect_regions_to_compact(
+                &self.tablet_registry,
+                region_ids,
+                compact_threshold,
+                &self.logger,
+            ) {
+                Ok(mut region_ids) => {
+                    for region_id in region_ids.drain(..) {
+                        let Some(mut tablet_cache) = self.tablet_registry.get(region_id) else {continue};
+                        let Some(tablet) = tablet_cache.latest() else {continue};
+                        for cf in &cf_names {
+                            if let Err(e) =
+                                tablet.compact_range_cf(cf, None, None, false, 1 /* threads */)
+                            {
+                                error!(
+                                    self.logger,
+                                    "compact range failed";
+                                    "regionid" => region_id,
+                                    "cf" => cf,
+                                    "err" => %e,
+                                );
+                            }
+                        }
+                        info!(
+                            self.logger,
+                            "compaction range finished";
+                            "region_id" => region_id,
+                        );
+                        fail_point!("raftstore-v2::CheckAndCompact::AfterCompact");
```

Comment:
In TiKV, we usually name failpoint in the style `[preposition_]funcion_name[_verb]`, `[ ]` means optional.